kotsadm/pkg/handlers/config.go

```diff
@@ -447,49 +447,52 @@ func decrypt(input string, cipher *crypto.AESCipher) (string, error) {
 }
 
 func getLaterVersions(versionedApp *apptypes.App, startSequence int64) (int64, []versiontypes.AppVersion, error) {
-	versions, err := version.GetVersions(versionedApp.ID)
+	thisAppVersion, err := store.GetStore().GetAppVersion(versionedApp.ID, startSequence)
 	if err != nil {
-		return -1, nil, errors.Wrap(err, "failed to get app versions")
+		return -1, nil, errors.Wrap(err, "failed to get this appversion")
 	}
 
-	thisUpdateCursor := -1
-	latestSequenceWithUpdateCursor := startSequence
-	laterVersions := map[int]versiontypes.AppVersion{}
-	for _, version := range versions {
-		if version.Sequence == startSequence {
-			thisUpdateCursor = version.UpdateCursor
-		}
-	}
-	if thisUpdateCursor == -1 {
-		err := fmt.Errorf("unable to find update cursor for sequence %d in %+v", startSequence, versions)
-		return -1, nil, err
+	laterAppVersions, err := store.GetStore().GetAppVersionsAfter(versionedApp.ID, startSequence)
+	if err != nil {
+		return -1, nil, errors.Wrap(err, "failed to get later app versions")
 	}
-	for _, version := range versions {
-		if version.UpdateCursor == thisUpdateCursor && version.Sequence > latestSequenceWithUpdateCursor {
-			latestSequenceWithUpdateCursor = version.Sequence
+
+	// latestSequenceWithThisUpdateCursor is the newest local version
+	// of the same upstream version
+	latestSequenceWithThisUpdateCursor := thisAppVersion.Sequence
+	for _, laterAppVersion := range laterAppVersions {
+		if laterAppVersion.KOTSKinds.Installation.Spec.UpdateCursor == thisAppVersion.KOTSKinds.Installation.Spec.UpdateCursor {
+			if laterAppVersion.Sequence > latestSequenceWithThisUpdateCursor {
+				latestSequenceWithThisUpdateCursor = laterAppVersion.Sequence
+			}
 		}
-		if version.UpdateCursor > thisUpdateCursor {
-			// save the latest sequence # for a given update cursor
-			ver, ok := laterVersions[version.UpdateCursor]
-			if !ok {
-				laterVersions[version.UpdateCursor] = version
-			} else if ver.Sequence < version.Sequence {
-				laterVersions[version.UpdateCursor] = version
+	}
+
+	laterVersions := map[string][]versiontypes.AppVersion{}
+	for _, laterAppVersion := range laterAppVersions {
+		if current, ok := laterVersions[laterAppVersion.KOTSKinds.Installation.Spec.UpdateCursor]; ok {
+			current = append(current, *laterAppVersion)
+			laterVersions[laterAppVersion.KOTSKinds.Installation.Spec.UpdateCursor] = current
+		} else {
+			laterVersions[laterAppVersion.KOTSKinds.Installation.Spec.UpdateCursor] = []versiontypes.AppVersion{
+				*laterAppVersion,
 			}
 		}
 	}
 
-	// ensure that the returned versions array is sorted by GVK
-	keys := []int{}
-	for key, _ := range laterVersions {
+	// ensure that the returned versions array is sorted
+	keys := []string{}
+	for key := range laterVersions {
 		keys = append(keys, key)
 	}
-	sort.Ints(keys)
+
+	// TODO sort by something in kotsutil that i need to write (these are either ints or semvers)
```

Comment:
maybe leftpad with 0s out to 10 or so?

so `0000000009` < `0000000010`

Reply:
i'm thinking that we just sort them as semvers?